tests/models/test_hooks.py

```diff
@@ -401,19 +420,30 @@ def test_trainer_model_hook_system_fit(tmpdir):
         'pytorch-lightning_version': __version__,
         'state_dict': ANY,
     }
+    if kwargs.get('amp_backend') == 'native':
+        saved_ckpt['native_amp_scaling_state'] = ANY
+    elif kwargs.get('amp_backend') == 'apex':
+        saved_ckpt['amp_scaling_state'] = ANY
+    device = torch.device('cuda:0' if 'gpus' in kwargs else 'cpu')
+
     expected = [
         dict(name='Callback.on_init_start', args=(trainer, )),
         dict(name='Callback.on_init_end', args=(trainer, )),
         dict(name='prepare_data'),
         dict(name='configure_callbacks'),
         dict(name='Callback.on_before_accelerator_backend_setup', args=(trainer, model)),
+        # FIXME
+        *([dict(name='train_dataloader')] if kwargs.get('plugins') == 'deepspeed' else []),
         dict(name='Callback.setup', args=(trainer, model), kwargs=dict(stage='fit')),
         dict(name='setup', kwargs=dict(stage='fit')),
         dict(name='configure_sharded_model'),
         dict(name='Callback.on_configure_sharded_model', args=(trainer, model)),
-        dict(name='configure_optimizers'),
+        # FIXME
+        *([dict(name='configure_optimizers')] if kwargs.get('plugins') != 'deepspeed' else []),
         dict(name='Callback.on_fit_start', args=(trainer, model)),
         dict(name='on_fit_start'),
+        # FIXME
+        *([dict(name='configure_optimizers')] if kwargs.get('plugins') == 'deepspeed' else []),
```

Comment:
@SeanNaren is the different order for deepspeed a bug or a requirement?

Reply:
Is as expected I think, probably has to do with the delaying logic found here: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/accelerator.py#L91

This is required to ensure the model is configured in `configure_sharded_model` before we initialize optimizers

Reply:
This is how it has been implemented as `init_optimizers` is a no-op: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/deepspeed.py#L485

We only call `configure_optimizers` if a user hasn't specified an optimizer or scheduler in the config object: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/deepspeed.py#L386

I think the comment should be expanded a bit because I think this is something that can be refactored in the future such that the call order isn't changed, but I haven't dug in depth here

Reply:
Is that because when you initially call `init_optimizers` the config is not available yet?

Would it be a bad idea to call `configure_optimizers` regardless and then overwrite the config if it doesn't have the optimizers?